### ISSUE TITLE
Integrating pytest in IDEs

### ISSUE THREAD

**Issue by fabioz** [pytest-dev/pytest#1613]
I'm the current PyDev maintainer (http://pydev.org), which provides an integration for running pytest tests directly inside of the IDE (which is probably a must for all IDEs out there, such as PyCharm, Wing, etc.), now, I must say I've done many versions of the integration and it's not really easy to get it right, so, I'd really appreciate if there was some guide to which would be the best way to do that integration (in a way that won't break as it evolves).

I know that pytest has the terminal as its main way of running tests, but as time goes by, I think many users are using IDEs to run it too and a guide/plugin implementation would be good for all implementors (ideally with test-cases inside pytest itself so that it doesn't break as time goes by).

If someone familiar with pytest decides to tackle this, what should be common for IDEs is something as:

```
Skip tests which the IDE didn't select to run

notify_tests_collected(tests_count)
notify_start_test(file, test)  # I.e.: 'c:/temp/test.py', 'TestCase.test1'
notify_test(when, status, stdout, stderr, file, test, time) # I.e.: 'call', 'ok', 'Standard Output', 'Standard Error with tracebacks', 'c:/temp/test.py', 'TestCase.test1', 0.2)
notify_test_run_finished(total_time)

Notes:
  - Should work with xdist/tox
  - Should report properly with xfail
  - Files must be given with the full path (changes to the cwd shouldn't mess with any report).
  - Outputs should always be present
  - Ideally should work with older versions of pytest too (it's hard to dictate what users should be using).
```

It seems simple, but using the pytest API, it's not really straightforward to get it right (although the API is pretty flexible, it's not straightforward to get that info in a consistent manner given different conditions and it's hard to know what should be private to pytest and what can be accessed without issues and how to convert from the pytest structure to that "simpler" info -- ideally pytest itself could have a plugin which converted to callbacks in that "simpler" structure).

As a reference, the current version in PyDev is: https://github.com/fabioz/Pydev/blob/master/plugins/org.python.pydev/pysrc/_pydev_runfiles/pydev_runfiles_pytest2.py

PyCharm has it at https://github.com/JetBrains/intellij-community/blob/fc0bf7c7a6e321b98f29f291ecfa12a7eb86bf63/python/helpers/pycharm/pytest_teamcity.py

Looking at the final code of both, while those implementations seem to work, I think the final code isn't great in any of those cases.

**Comment by nicoddemus** [comment]
Hey @fabioz,

Certainly it is on our interest to help IDE authors to integrate `py.test`.

But everything you listed seems pretty doable by writing [pytest hooks](http://pytest.org/latest/writing_plugins.html#pytest-hook-reference):

> Skip tests which the IDE didn't select to run
> notify_tests_collected(tests_count)

You can do both easily by implementing [pytest_collection_modifyitems](http://pytest.org/latest/writing_plugins.html#_pytest.hookspec.pytest_collection_modifyitems).

> notify_start_test(file, test)  # I.e.: 'c:/temp/test.py', 'TestCase.test1'

Implement either [pytest_runtest_protocol](http://pytest.org/latest/writing_plugins.html#_pytest.hookspec.pytest_runtest_protocol) or [pytest_runtest_setup](http://pytest.org/latest/writing_plugins.html#_pytest.hookspec.pytest_runtest_setup).

> notify_test(when, status, stdout, stderr, file, test, time) # I.e.: 'call', 'ok', 'Standard Output', 'Standard > Error with tracebacks', 'c:/temp/test.py', 'TestCase.test1', 0.2)

[pytest_runtest_logreport](http://pytest.org/latest/writing_plugins.html#_pytest.hookspec.pytest_runtest_logreport) provides all that.

> notify_test_run_finished(total_time)

There isn't a hook for this one specially, but I would use [pytest_unconfigure](http://pytest.org/latest/writing_plugins.html#_pytest.hookspec.pytest_unconfigure) and compute the total time myself.

Also, the hooks API is pretty stable and has guaranteed compatibility between releases.

**Comment by fabioz** [comment]
Hi @nicoddemus ;)

Yep, as I said, it works and that's what PyDev and PyCharm do right now -- and both have broken many times as pytest itself goes forward (you can take a look at the file history in the links I passed from github -- although the history from PyDev is probably not complete as the code was reorganized a few times).

If you take a look at both implementations, you can see both have to fiddle with things to get the full paths, deal with issues running from xdist (because hooks are called more times), force pytest to give the output, etc (so, although it appears straighforward, it really isn't).

I think I use most things you said already, but nevertheless, the integration has proven a bit brittle over time (although I did take chances on different APIs too, but maybe have probably chosen the wrong ones at those times -- because although the api seems stable, inner things which changed broke my integration, such as changes on the item, xfail, sections, etc -- although the hooks themselves are stable, the inner items which are worked with when dealing with them aren't as stable -- it's really a bit hard to know what's stable and what's not stable when dealing with those -- just to give an example, to get the full path, the item.location doesn't work, as it's relative, so, may break when making abspath(), so, I use item.fspath.strpath and save that instead, but it's hard to know if that's an API which will be maintained, even having a public access).

From your comment, I think I can use pytest_collection_modifyitems better for my current skipping thought, so, I'll improve that -- but still, what would be nicer is lowering the bar for that implementation and having a default base plugin code with simpler APIs ;)

Or providing a better documentation on what may be accessed or not at each of those hooks (again as an example, they deal with items, but it's hard to know what's stable on an item and how to make full paths without resorting to abspath which will depend on the current location).

**Comment by nicoddemus** [comment]
> because although the api seems stable, inner things which changed broke my integration, such as changes on the item, xfail, sections, etc -- although the hooks themselves are stable, the inner items which are worked with when dealing with them aren't as stable -- it's really a bit hard to know what's stable and what's not stable when dealing with those

Hmm that's unfortunate, I'm sure those were by accident. Do you remember which changes specifically affected the integration?

> just to give an example, to get the full path, the item.location doesn't work, as it's relative, so, may break when making abspath(), 

I'm curious, why can't you use `abspath(item.location[0])` in a hook implementation?

> so, I use item.fspath.strpath and save that instead, but it's hard to know if that's an API which will be maintained, even having a public access).

`item.fspath.strpath` should be stable, as I'm sure a lot of plugins must depend on it.

> Or providing a better documentation on what may be accessed or not at each of those hooks 

I assume everything listed in the [reference](http://pytest.org/latest/writing_plugins.html#reference-of-objects-involved-in-hooks) can be considered stable (and in fact did not change for a long time).

TBH I'm not sure how to tackle this issue, as there are no specific actionable items. Perhaps you could suggest specific new hooks or changes to existing ones that would help integration?

**Comment by fabioz** [comment]
I have an idea... How about pytest provides a plugin which converts those hooks to the simpler API I described (which is essentially what IDE implementors need to do)? That way it can be used as an authoritative source on how the integration should be done... 

My suggestion would be creating a module as:

ide_integration.py

which IDEs can use to add those simpler listeners (and that module itself would deal with the hooks in pytest and convert them to hooks for the IDE). It could be used as both, a simpler API and as a reference on how to implement it from scratch.

I can provide the initial code/tests for that if this is a sound suggestion.

> I'm curious, why can't you use abspath(item.location[0]) in a hook implementation?

As far as I know, pytest does not guarantee that the working directory is the same from when item.location[0] was created, so, doing that would yield the wrong result if the working dir changed in the meanwhile (am I wrong?). 

Ideally there'd be a way to force pytest to always make item.location[0] a full path so that the working dir is not an issue though ;)

**Comment by fabioz** [comment]
Ok, I started it to take a look at reviewing what I had and I have some questions I couldn't find by looking at the docs:

1. With xdist, what's the best way to know if a hook is being called in the main process or in the secondary process? (I'm currently experimenting, but I don't think there's any guarantee on how things will work from the docs)

2. How do I know which hooks are meant to be called in the main process and which hooks are meant to be called in the secondary process? (again, I'm experimenting, but there doesn't seem to be any guarantee it won't break in the future).

3. pytest_collection_modifyitems in xdist, seems to currently be called once for each node independently, but this seems strange... ideally it should be called only once in the main and nodes should already have that, but let's play with what pytest gives us now: given that pytest_collection_modifyitems is called twice in the nodes, how would I make a plugin which synchronizes that information from the multiple nodes into the master to give a callback which would say the number of items which were found?

**Comment by RonnyPfannschmidt** [comment]
1. slaves have `config.slaveinput`
2. its not formalized, but basically all collection and test running happens in the slaves
3. thats per design, the master has no test items, modify-items uses test items, it should also be possible to have different collection sets/context sets at the master

**Comment by fabioz** [comment]
@RonnyPfannschmidt thanks for the input... related to item 3, what do you feel would be the best way to synchronize that to make a report at the master node? Is there some message-passing mechanism I can use from the slave to talk to the master so that the master can do the reporting? I.e.: ideally there'd be some way to have a plugin which knows it'll run in x processes, gets the messages from those and makes a single report (right now, my code seems a bit hacky in that it's oblivious of that).

Also, I have another question: 4. when I have a report in pytest_runtest_logreport(report), what's the proper way to get the stdout/stderr for the test?

I currently have to setup my own redirection (because it seems that depending on how --capture is used, I'll either get the output or not), then I use the code below to get what may have been captured by pytest (depending on how the user configured, my redirection will get it or pytest will get it):

```
    for type_section, value in report.sections:
        if value:
            if type_section == 'stderr':
                error_contents += value
            else:
                captured_output += value

    if report.outcome != 'skipped':
        # On skipped, we'll have a traceback for the skip, which is not what we want.
        tw = TerminalWriter(stringio=True)
        tw.hasmarkup = False
        report.toterminal(tw)
        error_contents += tw.stringio.getvalue()
```

It seems that there's a report.longrepr, which is None or a failure representation from the docs, but I'm not sure how it should be used (is str(longrepr) the way to go?) -- it seems report.toterminal may be a better API to use -- is this a proper approach?

Also report.sections docs says "list of (secname, data) extra information which needs to marshallable", so, it's not clear if just appending to a string would actually always work.

**Comment by RonnyPfannschmidt** [comment]
reports do get serialized and logreport happens on master (alltho tracebacks get "mutilated")

you shouldn't mess with own capture while pytest capture is active, and xdist currently is impossible without capture

**Comment by fabioz** [comment]
Ok, so, I've talked to @nicoddemus about which things could be better on the integration and we ended up with the following set of improvements:

- Configure: ideally it should be possible to set an option to enable/disable capture in the config step (currently it seems options have a read-only configuration, but ideally a plug in should be able to change one of those). With this the PyDev plugin wouldn't need to do manual capturing out of pytest.

- xdist:
  * Hook to receive contents from remote in the master.
  * document that gateway.send could be used from the slave in a plugin to talk to the master in such a hook.
  * document better which default pytest hooks are expected to happen in the master/slave.
  * API to know if currently executing in slave or master.

- xfail: document better on how to detect fail/pass/skip on items (possibly an API better than hasattr(report, 'wasxfail') too? -- but without breaking the current 'informal' hasattr API).

- Provide some API to get the absolute path for items (currently things are relative to the collect working dir and changes to the current working dir could affect how the current relative path is resolved later on).

- ReprFileLocation: provide a way to customize the format and whether the absolute path should be used.

- Provide a way to get the original method name of items (i.e.: so that it's less 'fickle' to deal with parametrized tests).

- report: provide public API to get report contents (equivalent to report.toterminal, as the TerminalWriter doesn't seem to be part of the public API).

- report.sections: document what's in the keys (ideally provide constants for so that it's possible to identify an stderr/stdoutsection).

For new I'm reporting it here as a reference (as I think many items should be relatively fast to do, but if needed I can create a separate issue for each).

Note that the issues were raised after looking at https://github.com/fabioz/Pydev/blob/master/plugins/org.python.pydev/pysrc/_pydev_runfiles/pydev_runfiles_pytest2.py and checking things which could be better in the integration (but required things in pytest itself).

**Comment by nicoddemus** [comment]
Thanks @fabioz!

**Comment by nicoddemus** [comment]
Created #1790 and pytest-dev/pytest-xdist#76 to track those requirements better.

Closing this one now. :grin:
